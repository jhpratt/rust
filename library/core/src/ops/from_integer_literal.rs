#![allow(missing_docs)] // FIXME(jhpratt)

mod sealed {
    #[unstable(feature = "from_integer_literal", issue = "none")]
    pub trait Integer {}
    impl Integer for u8 {}
    impl Integer for u16 {}
    impl Integer for u32 {}
    impl Integer for u64 {}
    impl Integer for u128 {}
    impl Integer for usize {}
    impl Integer for i8 {}
    impl Integer for i16 {}
    impl Integer for i32 {}
    impl Integer for i64 {}
    impl Integer for i128 {}
    impl Integer for isize {}
}

// FIXME(jhpratt) Enforce all implementations are `const`.
#[cfg_attr(not(bootstrap), lang = "from_integer_literal")]
#[unstable(feature = "from_integer_literal", issue = "none")]
pub trait FromIntegerLiteral: Sized {
    // FIXME(jhpratt) Enforce the bound in the compiler?
    type Input: sealed::Integer;

    fn from_integer_literal(i: Self::Input) -> Self;
}
