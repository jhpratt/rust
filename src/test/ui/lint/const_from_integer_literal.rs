// compile-flags: --crate-type=lib

#![feature(from_integer_literal)]

use std::ops::FromIntegerLiteral;

struct Foo(u8);

impl FromIntegerLiteral for Foo {
    //~^ ERROR `FromIntegerLiteral` implementations must be `const`
    type Input = u8;

    fn from_integer_literal(n: Self::Input) -> Self {
        Foo(n)
    }
}
