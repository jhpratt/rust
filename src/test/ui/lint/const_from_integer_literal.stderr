error: `FromIntegerLiteral` implementations must be `const`
  --> $DIR/const_from_integer_literal.rs:9:1
   |
LL | / impl FromIntegerLiteral for Foo {
LL | |
LL | |     type Input = u8;
LL | |
...  |
LL | |     }
LL | | }
   | |_^
   |
   = note: `#[deny(const_from_integer_literal)]` on by default
   = note: this ensures that the value can be used in a const context

error: aborting due to previous error

